TESTING.md

```diff
@@ -0,0 +1,274 @@
+    beforeEach(() => {
+      // click the target button
+    });
+
+    it('should open the modal', () => {
+      // expect modal to be open
+    });
+  });
+});
+```
+
+The outline would look like:
+
+- Given the Default Modal is rendered
+  - when the target button is clicked
+    - then it should open the modal
+
+Tests should rely on aria attributes as much as possible for selecting and asserting. Test IDs
+should be used only to disambiguate DOM elements.
+
+Read [Intro to Cypress](https://docs.cypress.io/guides/core-concepts/introduction-to-cypress.html)
+for a good idea of how Cypress works. The core concept to Cypress is enqueued command chains that
+typically wrap jQuery collections. Cypress uses jQuery, Mocha, Chai, Sinon, Lodash and Moment. All
+are exposed to the developer of the test.
+
+Let's break that down. Cypress has a jQuery-like chainable API to select and interact with DOM
+nodes. Cypress control when and how often a command is run to ensure assertions either pass or time
+out. This is most confusing to those who approach Cypress commands as Promises.
+[Commands are not Promises](https://docs.cypress.io/guides/core-concepts/introduction-to-cypress.html#Commands-Are-Not-Promises).
+
+For example, imagine the following Cypress code:
+
+```ts
+cy.get('body')
+  .contains('button', 'Delete Item')
+  .click();
+
+cy.get('[data-testid="MyResult"]').should('contain', 'Success!');
+```
+
+The Cypress Command queue will show `['get', 'contains', 'click', 'get', 'should']`, but it will not
+run these commands immediately. There is also no need to use `.then` to wait for any previous
+condition to be met first. The Cypress run time will do the following:
+
+- Run `$('body')` every 50ms until `$('body')` returns a non-empty jQuery collection
+- Run a function that gets all `button` elements inside of the `body` element that have text content
+  that contains 'Delete Item' and that jQuery collection is not empty
+- Ensure the previous element is clickable (visible, not covered, has actual width/height), waiting
+  50ms to detect animations (waiting for animations to complete), etc. Basically make sure a user
+  could actually click on it.
+- Run `$('[data-testid="MyResult"]')` every 50ms until it returns a non-empty jQuery collection.
+  Usually clicking has side-effects like network requests to the server before the side-effect is
+  reflected in the DOM. Since Cypress implicitly retries, this command will _eventually_ return a
+  non-empty jQuery collection and the author doesn't need to explicitly do anything. Note this
+  selector _could_ return a non-empty jQuery collection instantly, but it might not contain
+  "Success!" yet...
+- Run internal `chai` + `chai-jquery` matchers (`expect($el).to.contain('Success!')`) until an error
+  is no longer thrown. `.should` will rerun the previous command _and_ the assertion until the
+  condition is met or a timeout occurs. This is the secret sauce to the stability of Cypress tests
```

Comment:
I think you use "time out" and "timeout" interchangeably in this piece, probably better to just pick one.

Reply:
Interesting. The only other place I could find "time out" was "...assertions either pass or time out." "timeout" is used as a verb here. Is that okay?